### PR TITLE
Add themes to help command when available #10318

### PR DESCRIPTION
**Pull request by DanyPDev**
# Description
The issue #10318 is resolved by introducing helper methods within the existing `get_documentation` function in the nu-engine crate. Initially, I considered using nu-color-config crate to convert HEX config color to ANSI color and employing the following method [https://github.com/nushell/nushell/blob/main/crates/nu-color-config/src/color_config.rs#L9C1-L20C2](https://github.com/nushell/nushell/blob/main/crates/nu-color-config/src/color_config.rs#L9C1-L20C2). However, this approach was deemed impractical due to circular dependencies. Consequently, in a manner akin to how we invoke the `table` command from the nu-command crate in `get_documentation` function  to create a themed-colored table, we invoke the `ansi` command from nu-command to obtain the ANSI theme color code.

# User-Facing Changes
Visual Changes Only: the help command now uses configured theme, else it falls back on default hard coded values.

# Tests + Formatting

# After Submitting

### CODE REVIEW

**Comment by DanyPDev** [comment]
Before:
image


After:
image

Theme used:
`let catppuccin_mocha = {
    rosewater: "#f5e0dc"
    flamingo: "#f2cdcd"
    pink: "#f5c2e7"
    mauve: "#cba6f7"
    red: "#f38ba8"
    maroon: "#eba0ac"
    peach: "#fab387"
    yellow: "#f9e2af"
    green: "#a6e3a1"
    teal: "#94e2d5"
    sky: "#89dceb"
    sapphire: "#74c7ec"
    blue: "#89b4fa"
    lavender: "#b4befe"
    text: "#cdd6f4"
    subtext1: "#bac2de"
    subtext0: "#a6adc8"
    overlay2: "#9399b2"
    overlay1: "#7f849c"
    overlay0: "#6c7086"
    surface2: "#585b70"
    surface1: "#45475a"
    surface0: "#313244"
    base: "#1e1e2e"
    mantle: "#181825"
    crust: "#11111b"
}

let catppuccin_theme = {
    separator: $catppuccin_mocha.overlay0
    # leading_trailing_space_bg: { attr: "n" }
    leading_trailing_space_bg: $catppuccin_mocha.overlay0
    header: { fg: $catppuccin_mocha.blue attr: "b" }
    empty: $catppuccin_mocha.lavender
    bool: $catppuccin_mocha.lavender
    int: $catppuccin_mocha.peach
    duration: $catppuccin_mocha.text
    filesize: {|size|
          if $size < 1mb {
            $catppuccin_mocha.green
        } else if $size < 100mb {
            $catppuccin_mocha.yellow
        } else if $size < 500mb {
            $catppuccin_mocha.peach
        } else if $size < 800mb {
            $catppuccin_mocha.maroon
        } else if $size > 800mb {
            $catppuccin_mocha.red
        }
    }
    date: {|| (date now) - $in |
        if $in < 1hr {
            $catppuccin_mocha.green
        } else if $in < 1day {
            $catppuccin_mocha.yellow
        } else if $in < 3day {
            $catppuccin_mocha.peach
        } else if $in < 1wk {
            $catppuccin_mocha.maroon
        } else if $in > 1wk {
            $catppuccin_mocha.red
        }
    }
    range: $catppuccin_mocha.text
    float: $catppuccin_mocha.text
    string: $catppuccin_mocha.text
    nothing: $catppuccin_mocha.text
    binary: $catppuccin_mocha.text
    cellpath: $catppuccin_mocha.text
    row_index: { fg: $catppuccin_mocha.mauve attr: "b" }
    record: $catppuccin_mocha.text
    list: $catppuccin_mocha.text
    block: $catppuccin_mocha.text
    hints: $catppuccin_mocha.overlay1
    search_result: { fg: $catppuccin_mocha.red bg: $catppuccin_mocha.text }
    shape_and: { fg: $catppuccin_mocha.pink attr: "b" }
    shape_binary: { fg: $catppuccin_mocha.pink attr: "b" }
    shape_block: { fg: $catppuccin_mocha.blue attr: "b" }
    shape_bool: $catppuccin_mocha.teal
    shape_custom: $catppuccin_mocha.green
    shape_datetime: { fg: $catppuccin_mocha.teal attr: "b" }
    shape_directory: $catppuccin_mocha.teal
    shape_external: $catppuccin_mocha.teal
    shape_externalarg: { fg: $catppuccin_mocha.green attr: "b" }
    shape_filepath: $catppuccin_mocha.teal
    shape_flag: { fg: $catppuccin_mocha.blue attr: "b" }
    shape_float: { fg: $catppuccin_mocha.pink attr: "b" }
    shape_garbage: { fg: $catppuccin_mocha.text bg: $catppuccin_mocha.red attr: "b" }
    shape_globpattern: { fg: $catppuccin_mocha.teal attr: "b" }
    shape_int: { fg: $catppuccin_mocha.pink attr: "b" }
    shape_internalcall: { fg: $catppuccin_mocha.teal attr: "b" }
    shape_list: { fg: $catppuccin_mocha.teal attr: "b" }
    shape_literal: $catppuccin_mocha.blue
    shape_match_pattern: $catppuccin_mocha.green
    shape_matching_brackets: { attr: "u" }
    shape_nothing: $catppuccin_mocha.teal
    shape_operator: $catppuccin_mocha.peach
    shape_or: { fg: $catppuccin_mocha.pink attr: "b" }
    shape_pipe: { fg: $catppuccin_mocha.pink attr: "b" }
    shape_range: { fg: $catppuccin_mocha.peach attr: "b" }
    shape_record: { fg: $catppuccin_mocha.teal attr: "b" }
    shape_redirection: { fg: $catppuccin_mocha.pink attr: "b" }
    shape_signature: { fg: $catppuccin_mocha.green attr: "b" }
    shape_string: $catppuccin_mocha.green
    shape_string_interpolation: { fg: $catppuccin_mocha.teal attr: "b" }
    shape_table: { fg: $catppuccin_mocha.blue attr: "b" }
    shape_variable: $catppuccin_mocha.pink

    background: $catppuccin_mocha.base
    foreground: $catppuccin_mocha.text
    cursor: $catppuccin_mocha.blue
}

$env.config.color_config = $catppuccin_theme
$env.LS_COLORS = (vivid generate catppuccin-mocha)`

File: crates/nu-engine/src/documentation.rs [hunk 1]
```diff
@@ -68,10 +70,13 @@ fn get_documentation(
...
-    const G: &str = "\x1b[32m"; // green
-    const C: &str = "\x1b[36m"; // cyan
-                                // was const BB: &str = "\x1b[1;34m"; // bold blue
-    const BB: &str = "\x1b[94m"; // light blue (nobold, should be bolding the *names*)
+    let g: String = get_ansi_color_for_component_or_default(engine_state, "header", "\x1b[32m"); // default: green
+
+    let c: String = get_ansi_color_for_component_or_default(engine_state, "row_index", "\x1b[36m"); // default: cyan
+                                                                                                    // was const bb: &str = "\x1b[1;34m"; // bold blue
+    let bb: String =
+        get_ansi_color_for_component_or_default(engine_state, "foreground", "\x1b[94m"); // default: light blue (nobold, should be bolding the *names*)
```
Comment (sholderbach) [inline crates/nu-engine/src/documentation.rs:78]: As they are now not just color short hands maybe it makes sense to expand those variables to more readable names.

File: crates/nu-engine/src/documentation.rs [hunk 2]
```diff
@@ -318,6 +323,97 @@ fn get_documentation(
...
+                    redirect_stderr: true,
+                    parser_info: HashMap::new(),
+                },
+                PipelineData::Empty,
+            ) {
+                if let Ok((str, ..)) = result.collect_string_strict(span) {
+                    return str;
+                }
+            }
+        }
```
Comment (sholderbach) [inline crates/nu-engine/src/documentation.rs:356]: As you noted a workaround. Maybe we can find another solution.

Using `ansi` as a command internally certainly carries some risk should something in the signature or behavior change, I am not sure if we properly test the help output with ansi.

But I think this also not a true stopper if we test it and don't see a catastrophic impact on performance.

File: crates/nu-engine/src/documentation.rs [hunk 3]
```diff
@@ -327,25 +423,40 @@ pub fn document_shape(shape: SyntaxShape) -> SyntaxShape {
...
     F: FnMut(&nu_protocol::Value) -> String,
 {
     //todo make these configurable -- pull from enginestate.config
-    const G: &str = "\x1b[32m"; // green
-    const C: &str = "\x1b[36m"; // cyan
-                                // was const BB: &str = "\x1b[1;34m"; // bold blue
-    const BB: &str = "\x1b[94m"; // light blue (nobold, should be bolding the *names*)
+    let g: String;
+    let c: String;
+    let bb: String;
```
Comment (sholderbach) [inline crates/nu-engine/src/documentation.rs:436]: Same as above.

File: crates/nu-engine/src/documentation.rs [hunk 4]
```diff
@@ -318,6 +323,97 @@ fn get_documentation(
...
+                }
+            }
+        }
+    }
+
+    default.to_string()
+}
+
+fn get_argument_for_color_value(
+    engine_state: &&EngineState,
```
Comment (sholderbach) [inline crates/nu-engine/src/documentation.rs:363]: This should probably remain a simple `&` and if necessary remove a level through dereference at the call site.
```suggestion
    engine_state: &EngineState,
```

File: crates/nu-engine/src/documentation.rs [hunk 5]
```diff
@@ -327,25 +423,40 @@ pub fn document_shape(shape: SyntaxShape) -> SyntaxShape {
 }
 
 pub fn get_flags_section<F>(
+    engine_state_opt: Option<&EngineState>,
```
Comment (sholderbach) [inline crates/nu-engine/src/documentation.rs:426]: Under which circumstances should this or has this to be `None`?

**Review by sholderbach — COMMENTED** [review]
Thank you for tackling this! 

Great quality of life improvement!

**Comment by fdncred** [comment]
Love your before and after screenshots, really shows a big difference. Thanks for working on this!

File: crates/nu-engine/src/documentation.rs (hunk 5 above)
Comment (DanyPDev) [reply crates/nu-engine/src/documentation.rs:436]: In the nu-plugin crate, `get_flags_section` gets called but there does not seem to be a way to get the engine state without creating one with all of its values [https://github.com/nushell/nushell/blob/main/crates/nu-plugin/src/plugin/mod.rs#L388](https://github.com/nushell/nushell/blob/main/crates/nu-plugin/src/plugin/mod.rs#L388).

**Comment by amtoine** [comment]
i love the screenshots and the results too :pray: 
can't wait to see beautiful help pages :scream:

File: crates/nu-engine/src/documentation.rs (hunk 2 above)
Comment (fdncred) [reply crates/nu-engine/src/documentation.rs:366]: I looked into this for a bit and couldn't see an easy way forward.

**Comment by DanyPDev** [comment]
I didn't realize the examples were colored in a different manner. For the "Examples" section, the color mapping happens when get_documentation calls nu-highlight command. The mapping can be found here [https://github.com/nushell/nushell/blob/main/crates/nu-parser/src/flatten.rs#L50-L83](https://github.com/nushell/nushell/blob/main/crates/nu-parser/src/flatten.rs#L50-L83). To give visibility on how things are mapped, see the image below.
image

**Comment by DanyPDev** [comment]
Commited changes for the help color to shape_string, shape_external and shape_block. Still looks pretty neat 😄!
image

File: crates/nu-engine/src/documentation.rs (hunk 5 above)
Comment (sholderbach) [reply crates/nu-engine/src/documentation.rs:432]: Ah that is unfortunate, but yeah simplifying the plugin side of things is probably too much for this PR.
We can keep that as a todo for the future.

**Review by sholderbach — APPROVED** [review]
Thanks for pushing the improvements!

Torn with whether we should add additional config points. I like the default experience you provide with those.

**Comment by fdncred** [comment]
Let's move forward with this, if we need to continue tweaking we can. Thanks!

**Pull request merged**